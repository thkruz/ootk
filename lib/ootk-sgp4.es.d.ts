/**
 * @author Theodore Kruczek.
 * @description Orbital Object ToolKit (OOTK) is a collection of tools for working
 * with satellites and other orbital objects. The Sgp4 module contains a TypeScript
 * port of the 2020 version of sgp4unit.cpp from "Fundamentals of Astrodynamics and
 * Applications" by David Vallado. All of the original comments and notes are
 * inserted in the code below in order provide context to the functions and clarify
 * any adjustments made for TypeScript/JavaScript compatibility.
 *
 * @copyright MIT License.
 * Copyright (c) 2020-2021 Theodore Kruczek
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/** @interface SatelliteRecord contains all of the orbital parameters necessary for running SGP4. It is generated by Sgp4.createSatrec.*/
interface SatelliteRecord {
    a: number;
    am: number;
    alta: number;
    altp: number;
    argpdot: number;
    argpo: number;
    aycof: number;
    bstar: number;
    cc1: number;
    cc4: number;
    cc5: number;
    con41: number;
    d2: number;
    d3: number;
    d4: number;
    d5232: number;
    d5421: number;
    d5433: number;
    dedt: number;
    delmo: number;
    del1: number;
    ecco: number;
    em: number;
    epochdays: number;
    epochyr: number;
    error: number;
    eta: number;
    gsto: number;
    im: number;
    inclo: number;
    init: string;
    isimp: number;
    jdsatepoch: number;
    mdot: number;
    method: string;
    mo: number;
    mm: number;
    nddot: number;
    ndot: number;
    no: number;
    nodecf: number;
    nodedot: number;
    nodeo: number;
    om: number;
    Om: number;
    omgcof: number;
    operationmode: string;
    satnum: string;
    sinmao: number;
    t: number;
    t2cof: number;
    t3cof: number;
    t4cof: number;
    t5cof: number;
    x1mth2: number;
    x7thm1: number;
    xlcof: number;
    xmcof: number;
    xfact: number;
    xlamo: number;
    xli: number;
    xgh4: number;
    xgh3: number;
    xh2: number;
    xi2: number;
    xi3: number;
    xl2: number;
    xl3: number;
    xl4: number;
    zmol: number;
    zmos: number;
    dmdt: number;
    dnodt: number;
    domdt: number;
    e3: number;
    ee2: number;
    peo: number;
    pgho: number;
    pho: number;
    PInco: number;
    plo: number;
    se2: number;
    se3: number;
    sgh2: number;
    sgh3: number;
    sgh4: number;
    sh2: number;
    sh3: number;
    si2: number;
    si3: number;
    sl2: number;
    sl3: number;
    sl4: number;
    xgh2: number;
    xh3: number;
    tumin: number;
    radiusearthkm: number;
    irez: number;
    d3210: number;
    d3222: number;
    d4410: number;
    d4422: number;
    d5220: number;
    del2: number;
    del3: number;
    didt: number;
    atime: number;
    j2: number;
    j3: number;
    j4: number;
    mus: number;
    xke: number;
    j3oj2: number;
    xni: number;
    d2201: number;
    d2211: number;
    nm: number;
}
/** @typedef StateVector is a three dimensional position and velocity vector output from Sgp4.propagate.*/
declare type StateVector = {
    position: {
        x: number;
        y: number;
        z: number;
    } | boolean;
    velocity: {
        x: number;
        y: number;
        z: number;
    } | boolean;
};
/** @typedef vec3 is a flat three dimensional array used in vector math.*/
declare type vec3 = [number, number, number];
declare class Sgp4 {
    private static dpper;
    private static dscom;
    private static dsinit;
    private static dspace;
    private static initl;
    private static sgp4init;
    static propagate(satrec: SatelliteRecord, tsince: number): StateVector;
    private static getgravconst;
    static createSatrec(tleLine1: string, tleLine2: string, whichconst?: string, opsmode?: string): SatelliteRecord;
    static gstime(jdut1: number): number;
    static sgn(x: number): number;
    static mag(x: vec3): number;
    static cross(vec1: vec3, vec2: vec3): vec3;
    static dot(x: vec3, y: vec3): number;
    static angle(vec1: vec3, vec2: vec3): number;
    static asinh(xval: number): number;
    static newtonnu(ecc: number, nu: number): {
        e0: number;
        m: number;
    };
    static sinh(x: number): number;
    static rv2coe(r: vec3, v: vec3, mus: number): {
        p: number;
        a: number;
        ecc: number;
        incl: number;
        omega: number;
        argp: number;
        nu: number;
        m: number;
        arglat: number;
        truelon: number;
        lonper: number;
    };
    static jday(year: number, mon: number, day: number, hr: number, minute: number, sec: number): {
        jd: number;
        jdFrac: number;
    };
    static days2mdhms(year: number, days: number): {
        mon: number;
        day: number;
        hr: number;
        minute: number;
        sec: number;
    };
    static invjday(jd: number, jdfrac: number): {
        year: number;
        mon: number;
        day: number;
        hr: number;
        minute: number;
        sec: number;
    };
}
export { Sgp4 };
