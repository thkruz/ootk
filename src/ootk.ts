/**
 * @author @thkruz Theodore Kruczek
 * @description Orbital Object ToolKit (OOTK) is a collection of tools for working
 * with satellites and other orbital objects.
 *
 * @license MIT License
 * @Copyright (c) 2020-2024 Theodore Kruczek
 *
 * Many of the classes are based off of the work of @david-rc-dayton and his
 * Pious Squid library (https://github.com/david-rc-dayton/pious_squid).
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

export { Celestial } from './body/Celestial';
export { Earth } from './body/Earth';
export { Moon } from './body/Moon';
export { Sun } from './body/Sun';
export * from './objects/index';
export * from './operations/operations';
export { Sgp4 } from './sgp4/sgp4';
export * from './time/time';
export { Tle } from './tle/tle';
export * from './transforms';
export * from './types/types';
export * from './utils/constants';
export { Utils } from './utils/utils';
