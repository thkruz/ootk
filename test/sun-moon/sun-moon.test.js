/**
 * @file   Test Suite to verify tle functions work as expected
 * @author Theodore Kruczek.
 * @since  1.3.0
 */

import { Utils } from '../../lib/ootk';

const { SunMath, MoonMath } = Utils;
// Use number of milliseconds since epoch instead of local year, month, day, etc for consistency across machines
const dateObj = new Date(1661400000000);

describe('Sun and Moon', () => {
  test('SunMath Unit Tests', () => {
    const d = SunMath.julian2date(SunMath.date2jSince2000(dateObj));
    const c = SunMath.getSunRaDec(d);

    SunMath.getStarAzEl(dateObj, 0, 0, c.ra, c.dec);
    SunMath.getSunAzEl(dateObj, 0, 0);
  });

  test('Local Solar Time', () => {
    // Use number of milliseconds since epoch instead of local year, month, day, etc for consistency across machines
    const lst = SunMath.getSolarTime(new Date(1658807880000), -5, -71);

    expect(lst.toUTCString()).toEqual('Tue, 26 Jul 2022 04:07:49 GMT');
  });

  test('MoonMath Unit Tests', () => {
    expect(MoonMath.getMoonIllumination(dateObj)).toMatchSnapshot();

    MoonMath.getMoonPosition(dateObj, 0, 0);
    MoonMath.getMoonTimes(dateObj, 0, 0, true);
    MoonMath.getMoonTimes(dateObj, -10, -10, false);
  });

  test('getMoonIllumination returns fraction and angle of moons illuminated limb and phase', () => {
    const moonIllum = MoonMath.getMoonIllumination(dateObj);

    expect(moonIllum).toMatchSnapshot();
  });
});

/**
 * @author Robert Myers @xqjibz
 * See: https://github.com/mourner/suncalc/pull/35
 */
describe('Test for #6 fix', () => {
  test('Test for #6 fix', () => {
    const date = new Date('2013-03-05UTC');
    const result = (SunMath.getSunAzEl(date, 50.5, 30.5).az * 180) / Math.PI;

    expect(result).toBeCloseTo(36.742354609606814);
  });
});

describe('Test for variety of date/time stamps', () => {
  describe('getTimes day detection works with a variety of date times', () => {
    const lat = 47.606209;
    const lng = -122.332069;
    // const testDateDay = 4;
    const testDateStrings = [
      'Mon, 04 Mar 2013 00:00:01 UTC+2',
      'Mon, 04 Mar 2013 00:00:01 PDT',
      'Mon, 04 Mar 2013 00:00:01 PST',
      'Mon, 04 Mar 2013 00:00:01 EDT',
      'Mon, 04 Mar 2013 00:00:01 EST',
      'Mon, 04 Mar 2013 00:00:01 UTC',
      'Mon, 04 Mar 2013 12:00:00 PDT',
      'Mon, 04 Mar 2013 12:00:00 PST',
      'Mon, 04 Mar 2013 12:00:00 EDT',
      'Mon, 04 Mar 2013 12:00:00 EST',
      'Mon, 04 Mar 2013 12:00:00 UTC',
      'Mon, 04 Mar 2013 23:59:59 PDT',
      'Mon, 04 Mar 2013 23:59:59 PST',
      'Mon, 04 Mar 2013 23:59:59 EDT',
      'Mon, 04 Mar 2013 23:59:59 EST',
      'Mon, 04 Mar 2013 23:59:59 UTC',
    ];

    for (let i = 0, l = testDateStrings.length; i < l; i++) {
      test(`${testDateStrings[i]}`, () => {
        const date = new Date(testDateStrings[i]);
        const testDateDay = date.getDate();

        const times = SunMath.getTimes(date, lat, lng);

        expect(times.solarNoon.getDate()).toEqual(testDateDay);
      });
    }
  });
});

describe('Suncalc.js tests', () => {
  const date = new Date('2013-03-05 UTC');
  const lat = 50.5;
  const lon = 30.5;
  const alt = 2000;

  const testTimes = {
    solarNoon: '2013-03-05T10:10:57Z',
    nadir: '2013-03-05T22:10:57Z',
    sunriseStart: '2013-03-05T04:34:56Z',
    sunsetEnd: '2013-03-05T15:46:57Z',
    sunriseEnd: '2013-03-05T04:38:19Z',
    sunsetStart: '2013-03-05T15:43:34Z',
    civilDawn: '2013-03-05T04:02:17Z',
    civilDusk: '2013-03-05T16:19:36Z',
    nauticalDawn: '2013-03-05T03:24:31Z',
    nauticalDusk: '2013-03-05T16:57:22Z',
    astronomicalDawn: '2013-03-05T02:46:17Z',
    astronomicalDusk: '2013-03-05T17:35:36Z',
    goldenHourDawnEnd: '2013-03-05T05:19:01Z',
    goldenHourDuskStart: '2013-03-05T15:02:52Z',
  };

  const heightTestTimes = {
    solarNoon: '2013-03-05T10:10:57Z',
    nadir: '2013-03-05T22:10:57Z',
    sunriseStart: '2013-03-05T04:25:07Z',
    sunsetEnd: '2013-03-05T15:56:46Z',
  };

  test('getTimes returns sun phases for the given date and location', () => {
    const times = SunMath.getTimes(date, lat, lon, 0, true);

    // eslint-disable-next-line guard-for-in
    for (const i in testTimes) {
      expect(times[i].toUTCString()).toEqual(new Date(testTimes[i]).toUTCString());
    }
  });

  test('getTimes adjusts sun phases when additionally given the observer height', () => {
    const times = SunMath.getTimes(date, lat, lon, alt, true);

    // eslint-disable-next-line guard-for-in
    for (const i in heightTestTimes) {
      expect(times[i].toUTCString()).toEqual(new Date(heightTestTimes[i]).toUTCString());
    }
  });

  test('getSunAzEl returns azimuth and altitude for the given time and location', () => {
    const sunPos = SunMath.getSunAzEl(date, lat, lon);

    expect(sunPos.az).toBeCloseTo(0.6412750628729547);
    expect(sunPos.el).toBeCloseTo(-0.7000406838781611);
  });

  test('getMoonIllumination returns fraction and angle of moons illuminated limb and phase', () => {
    const moonIllum = MoonMath.getMoonIllumination(date);

    expect(moonIllum.fraction).toBeCloseTo(0.4848068202456373);
    expect(moonIllum.phaseValue).toBeCloseTo(0.7548368838538762);
    expect(moonIllum.angle).toBeCloseTo(1.6732942678578346);
  });

  test('getMoonPosition returns moon position data given time and location', () => {
    const moonPos = MoonMath.getMoonPosition(date, lat, lon);

    expect(moonPos.az).toBeCloseTo(2.1631927013459706);
    expect(moonPos.el).toBeCloseTo(0.014551482243892251);
    expect(moonPos.rng).toBeCloseTo(364121.37256256194);
  });

  test('getMoonTimes returns moon rise and set times', () => {
    const moonTimes = MoonMath.getMoonTimes(new Date('2013-03-04UTC'), lat, lon, true);

    expect(moonTimes.rise.toUTCString()).toEqual('Mon, 04 Mar 2013 23:54:29 GMT');
    expect(moonTimes.set.toUTCString()).toEqual('Mon, 04 Mar 2013 07:47:58 GMT');
  });
});

describe('Tests from Hypnos3', () => {
  const date = new Date('2013-03-05UTC');
  const lat = -34.0;
  const lon = 151.0;

  const testTimes = {
    // southern hemisphere
    solarNoon: '2013-03-05T02:09:01.832Z',
    nadir: '2013-03-05T14:09:01.832Z',
    goldenHourDuskStart: '2013-03-05T07:56:33.416Z',
    goldenHourDawnEnd: '2013-03-04T20:21:30.248Z',
    sunsetStart: '2013-03-05T08:27:05.997Z',
    sunriseEnd: '2013-03-04T19:50:57.667Z',
    sunsetEnd: '2013-03-05T08:29:41.731Z',
    sunriseStart: '2013-03-04T19:48:21.933Z',
    goldenHourDuskEnd: '2013-03-05T08:30:30.554Z',
    goldenHourDawnStart: '2013-03-04T19:47:33.110Z',
    blueHourDuskStart: '2013-03-05T08:45:10.179Z',
    blueHourDawnEnd: '2013-03-04T19:32:53.485Z',
    civilDusk: '2013-03-05T08:54:59.722Z',
    civilDawn: '2013-03-04T19:23:03.942Z',
    blueHourDuskEnd: '2013-03-05T09:04:52.263Z',
    blueHourDawnStart: '2013-03-04T19:13:11.401Z',
    nauticalDusk: '2013-03-05T09:24:48.289Z',
    nauticalDawn: '2013-03-04T18:53:15.375Z',
    amateurDusk: '2013-03-05T09:39:57.120Z',
    amateurDawn: '2013-03-04T18:38:06.544Z',
    astronomicalDusk: '2013-03-05T09:55:18.657Z',
    astronomicalDawn: '2013-03-04T18:22:45.007Z',
  };

  test('southern hemisphere', () => {
    const times = SunMath.getTimes(date, lat, lon);

    // eslint-disable-next-line guard-for-in
    for (const i in testTimes) {
      expect(times[i].toUTCString()).toEqual(new Date(testTimes[i]).toUTCString());
    }
  });

  test('getSolarTime returns the solar time', () => {
    const solarTime = SunMath.getSolarTime(date, -5, -71);

    expect(solarTime.toUTCString()).toEqual('Tue, 05 Mar 2013 00:03:33 GMT');
  });
});
