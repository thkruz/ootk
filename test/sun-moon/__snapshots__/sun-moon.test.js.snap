// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sun and Moon MoonMath Unit Tests 1`] = `
Object {
  "angle": 2.0187380462851525,
  "fraction": 0.04918501150967336,
  "next": Object {
    "date": "2022-08-27T06:46:57.840Z",
    "firstQuarter": Object {
      "date": "2022-09-03T15:57:58.534Z",
      "value": 1662220678534.5,
    },
    "fullMoon": Object {
      "date": "2022-09-11T01:08:59.229Z",
      "value": 1662858539229,
    },
    "newMoon": Object {
      "date": "2022-08-27T06:46:57.840Z",
      "value": 1661582817840,
    },
    "thirdQuarter": Object {
      "date": "2022-09-18T10:19:59.923Z",
      "value": 1663496399923.5,
    },
    "type": "newMoon",
    "value": 1661582817840,
  },
  "phase": Object {
    "code": ":waning_crescent_moon:",
    "css": "wi-moon-wan-cres",
    "emoji": "ðŸŒ˜",
    "from": 0.783863193308711,
    "id": "waningCrescentMoon",
    "name": "Waning Crescent",
    "to": 0.966136806691289,
    "weight": 6.3825,
  },
  "phaseValue": 0.9288143180691704,
}
`;

exports[`Sun and Moon getMoonIllumination returns fraction and angle of moons illuminated limb and phase 1`] = `
Object {
  "angle": 2.0187380462851525,
  "fraction": 0.04918501150967336,
  "next": Object {
    "date": "2022-08-27T06:46:57.840Z",
    "firstQuarter": Object {
      "date": "2022-09-03T15:57:58.534Z",
      "value": 1662220678534.5,
    },
    "fullMoon": Object {
      "date": "2022-09-11T01:08:59.229Z",
      "value": 1662858539229,
    },
    "newMoon": Object {
      "date": "2022-08-27T06:46:57.840Z",
      "value": 1661582817840,
    },
    "thirdQuarter": Object {
      "date": "2022-09-18T10:19:59.923Z",
      "value": 1663496399923.5,
    },
    "type": "newMoon",
    "value": 1661582817840,
  },
  "phase": Object {
    "code": ":waning_crescent_moon:",
    "css": "wi-moon-wan-cres",
    "emoji": "ðŸŒ˜",
    "from": 0.783863193308711,
    "id": "waningCrescentMoon",
    "name": "Waning Crescent",
    "to": 0.966136806691289,
    "weight": 6.3825,
  },
  "phaseValue": 0.9288143180691704,
}
`;
